ingestor/clickhouse/uploader.go

```diff
@@ -0,0 +1,655 @@
+				}
+			}
+		}
+		return stats.error(err, retryable)
+	}
+
+	var (
+		writer       batchWriter
+		schemaDef    Schema
+		schemaLoaded bool
+		schemaID     string
+		headerSig    string
+		rowsInBatch  int
+		processed    int
+		flushTimer   *time.Timer
+	)
+
+	defer func() {
+		if flushTimer != nil {
+			flushTimer.Stop()
+		}
+		if writer != nil {
+			_ = writer.Close()
+		}
+	}()
+
+	flushInterval := u.cfg.Batch.FlushInterval
+
+	for _, segment := range batch.Segments {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		sr, err := wal.NewSegmentReader(segment.Path)
+		if errors.Is(err, os.ErrNotExist) {
+			continue
+		}
+		if err != nil {
+			if u.log != nil {
+				u.log.Error("failed to open segment", slog.String("path", segment.Path), slog.String("error", err.Error()))
+			}
+			return handleError(err, false)
+		}
+
+		processed++
+		stats.addBytes(segment.Size)
+
+		bufReader := bufio.NewReader(sr)
+		header, err := readHeader(bufReader)
+		if err != nil {
+			sr.Close()
+			if u.log != nil {
+				u.log.Error("failed to read segment header", slog.String("path", segment.Path), slog.String("error", err.Error()))
+			}
+			return handleError(err, false)
+		}
+
+		if headerSig == "" {
+			headerSig = header
+			if u.log != nil {
+				u.log.Info("loaded WAL segment header",
+					slog.String("table", batch.Table),
+					slog.String("path", segment.Path),
+					slog.String("header_signature", header))
+			}
+			_, _, schemaName, _, err := wal.ParseFilename(segment.Path)
+			if err != nil {
+				sr.Close()
+				if u.log != nil {
+					u.log.Error("failed to parse segment filename", slog.String("path", segment.Path), slog.String("error", err.Error()))
+				}
+				return handleError(err, false)
+			}
+
+			schemaID = schemaName
+			mapping, err := schema.UnmarshalSchema(header)
+			if err != nil {
+				sr.Close()
+				if u.log != nil {
+					u.log.Error("failed to unmarshal schema", slog.String("path", segment.Path), slog.String("error", err.Error()))
+				}
+				return handleError(err, false)
+			}
+			if len(mapping) == 0 {
+				sr.Close()
+				return handleError(fmt.Errorf("empty schema for table %s", batch.Table), false)
+			}
+
+			schemaDef, err = u.ensureSchema(ctx, batch.Table, schemaID, mapping)
+			if err != nil {
+				sr.Close()
+				if u.log != nil {
+					u.log.Error("failed to ensure schema", slog.String("table", batch.Table), slog.String("error", err.Error()))
+				}
+				return handleError(err, true)
+			}
+
+			if u.log != nil {
+				columnSpecs := make([]string, 0, len(schemaDef.Columns))
+				for _, col := range schemaDef.Columns {
+					columnSpecs = append(columnSpecs, fmt.Sprintf("%s:%s", col.Name, col.Type))
+				}
+				u.log.Info("prepared ClickHouse schema",
+					slog.String("table", batch.Table),
+					slog.String("schema_id", schemaID),
+					slog.Any("columns", columnSpecs))
+			}
+
+			writer, err = u.conn.PrepareInsert(ctx, batch.Database, batch.Table, schemaDef.Columns)
```

Comment:
I know you plan to revamp this bit, but you might be able to just import the segment directly like we do with adx size and:

INSERT INTO sometable
FROM INFILE 'data_small.csv'
FORMAT CSV

You'd need to be able to stream the segments via the statement though.